### PR TITLE
Add abstraction layer for the validation of transactions

### PR DESCRIPTION
**Pull request by clement-fischer**

### CODE REVIEW

File: src/libValidator/Validator.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+#include <vector>
+
+#include "Validator.h"
+#include "libData/AccountData/Account.h"
+#include "libMediator/Mediator.h"
+
+using namespace std;
+using namespace boost::multiprecision;
+
+ZilliqaValidator::ZilliqaValidator() { m_mediator = nullptr; }
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: Class name should be just Validator

File: src/libValidator/Validator.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        lock_guard<mutex> g(m_mutexTxnNonceMap);
+        m_txnNonceMap.clear();
+    }
+}
+
+#ifndef IS_LOOKUP_NODE
+bool Validator::CheckCreatedTransaction(const Transaction& tx) const
+{
+    LOG_MARKER();
+
```
Comment (ckyang) [inline src/libValidator/Validator.cpp]: Try to check nullptr for m_mediator

File: src/libValidator/Validator.cpp [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        return false;
+    }
+
+    return true;
+}
+
+bool Validator::CheckCreatedTransactionFromLookup(const Transaction& tx)
+{
+    LOG_MARKER();
+
```
Comment (ckyang) [inline src/libValidator/Validator.cpp]: Try to check nullptr for m_mediator

File: src/libValidator/Validator.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+* and, to the extent permitted by law, all liability for your use of the code is disclaimed. 
+* Some programs in this code are governed by the GNU General Public License v3.0 (available at 
+* https://www.gnu.org/licenses/gpl-3.0.en.html) (‘GPLv3’). The programs that are governed by 
+* GPLv3.0 are those programs that are located in the folders src/depends and tests/depends 
+* and which include a reference to GPLv3 in their program files.
+**/
+
+#include <vector>
+
+#include "Validator.h"
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: alpbaetical order

File: src/libValidator/Validator.cpp [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    // Check if from account is sharded here
+    const PubKey& senderPubKey = tx.GetSenderPubKey();
+    Address fromAddr = Account::GetAddressFromPublicKey(senderPubKey);
+    unsigned int shardID = m_mediator->m_node->getShardID();
+    unsigned int numShards = m_mediator->m_node->getNumShards();
+    unsigned int correct_shard
+        = Transaction::GetShardIndex(fromAddr, numShards);
+    if (correct_shard != shardID)
+    {
+        LOG_EPOCH(WARNING, to_string(m_mediator->m_currentEpochNum).c_str(),
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: combine all these into one log_epoch

File: src/libValidator/Validator.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    {
+        LOG_GENERAL(INFO, "New account is added: " << toAddr);
+        AccountStore::GetInstance().AddAccount(
+            toAddr, {0, 0, dev::h256(), dev::h256()});
+    }
+
+    // Check if transaction amount is valid
+    if (AccountStore::GetInstance().GetBalance(fromAddr) < tx.GetAmount())
+    {
+        LOG_EPOCH(WARNING, to_string(m_mediator->m_currentEpochNum).c_str(),
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: combine all these into one log_epoch

File: src/libValidator/Validator.cpp [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    // Check if from account is sharded here
+    const PubKey& senderPubKey = tx.GetSenderPubKey();
+    Address fromAddr = Account::GetAddressFromPublicKey(senderPubKey);
+    unsigned int shardID = m_mediator->m_node->getShardID();
+    unsigned int numShards = m_mediator->m_node->getNumShards();
+    unsigned int correct_shard
+        = Transaction::GetShardIndex(fromAddr, numShards);
+    if (correct_shard != shardID)
+    {
+        LOG_EPOCH(WARNING, to_string(m_mediator->m_currentEpochNum).c_str(),
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: combine all these into one log_epoch

File: src/libValidator/Validator.cpp [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    return true;
+}
+
+bool Validator::CheckCreatedTransactionFromLookup(const Transaction& tx)
+{
+    LOG_MARKER();
+
+    if (m_mediator == nullptr)
+    {
+        throw std::runtime_error("Pointer to Mediator is nullptr");
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: log the error but don't make the program terminate

File: src/libValidator/Validator.cpp [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+                  "This shard    = " << m_mediator->m_node->getShardID());
+        // // Transaction created from the GenTransactionBulk will be rejected
+        // // by all shards but one. Next line is commeted to avoid this
+        // return false;
+    }
+
+    // Check if from account exists in local storage
+    if (!AccountStore::GetInstance().DoesAccountExist(fromAddr))
+    {
+        LOG_GENERAL(INFO,
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: use warning

File: src/libValidator/Validator.cpp [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+            m_txnNonceMap.insert(make_pair(fromAddr, tx.GetNonce()));
+        }
+        else
+        {
+            if (tx.GetNonce() != m_txnNonceMap.at(fromAddr) + 1)
+            {
+                LOG_EPOCH(WARNING,
+                          to_string(m_mediator->m_currentEpochNum).c_str(),
+                          "Tx nonce not in line with account state!");
+                LOG_EPOCH(INFO,
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: combine all these into one log_epoch

File: src/libValidator/Validator.cpp [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+        AccountStore::GetInstance().AddAccount(
+            toAddr, {0, 0, dev::h256(), dev::h256()});
+    }
+
+    // Check if transaction amount is valid
+    if (AccountStore::GetInstance().GetBalance(fromAddr) < tx.GetAmount())
+    {
+        LOG_EPOCH(WARNING, to_string(m_mediator->m_currentEpochNum).c_str(),
+                  "Insufficient funds in source account!");
+        LOG_EPOCH(INFO, to_string(m_mediator->m_currentEpochNum).c_str(),
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: combine all these into one log_epoch

File: src/libValidator/Validator.cpp [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+        {
+            LOG_GENERAL(INFO, "Txn from " << fromAddr << "is new.");
+
+            if (tx.GetNonce()
+                != AccountStore::GetInstance().GetNonce(fromAddr) + 1)
+            {
+                LOG_EPOCH(WARNING,
+                          to_string(m_mediator->m_currentEpochNum).c_str(),
+                          "Tx nonce not in line with account state!");
+                LOG_EPOCH(INFO,
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: combine into one log_epoch

File: src/libValidator/Validator.cpp [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+using namespace boost::multiprecision;
+
+Validator::Validator(Mediator& mediator)
+    : m_mediator(&mediator)
+{
+}
+
+Validator::~Validator() {}
+
+bool Validator::verifyTransaction(const Transaction& tran) const
```
Comment (bb111189) [inline src/libValidator/Validator.cpp]: Upper case for first letter in function

File: src/libValidator/Validator.h [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+    // Nonce information
+    std::mutex m_mutexTxnNonceMap;
+    std::unordered_map<Address, boost::multiprecision::uint256_t> m_txnNonceMap;
+
+public:
+    Validator(Mediator& mediator);
+    ~Validator();
+    std::string name() const override { return "Validator"; }
+    bool verifyTransaction(const Transaction& tran) const override;
```
Comment (bb111189) [inline src/libValidator/Validator.h]: Upper case for first letter of function name

File: src/libValidator/Validator.h [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+class Mediator;
+
+class ValidatorBase
+{
+public:
+    virtual ~ValidatorBase() {}
+    virtual std::string name() const = 0;
+
+    /// Verifies the transaction w.r.t given pubKey and signature
+    virtual bool verifyTransaction(const Transaction& tran) const = 0;
```
Comment (bb111189) [inline src/libValidator/Validator.h]: Upper case for first letter of function name

File: src/libValidator/Validator.cpp (hunk 4 above)
Comment (clement-fischer) [reply src/libValidator/Validator.cpp]: I am not sure what you mean here.
It seems like in most files (all the ones I checked at least), includes are sorted in the following order:
1. Standard headers (alphabetical order)
2. Class header (matching .h file for a .cpp file)
3. Project headers (alphabetical order)

It is also what is is written in the wiki. Can you please clarify which convention you would like to follow ?

File: src/libValidator/Validator.cpp (hunk 5 above)
Comment (bb111189) [reply src/libValidator/Validator.cpp]: not yet resolved

File: src/libValidator/Validator.cpp (hunk 6 above)
Comment (bb111189) [reply src/libValidator/Validator.cpp]: not yet resolved

File: src/libValidator/Validator.cpp (hunk 7 above)
Comment (bb111189) [reply src/libValidator/Validator.cpp]: not yet resolved

**Comment by clement-fischer** [comment]
Well then I don't know what you want. Initially you highlighted 4 lines to be combined, now they are. By the way this is not part of the refactoring at all.
________________________________

File: tests/Data/Test_Transaction.cpp [hunk 16]
```diff
@@ -42,6 +43,9 @@ BOOST_AUTO_TEST_CASE(test1)
 
     Address toAddr;
 
+    Mediator* m = nullptr;
+    ValidatorBase* m_validator = new Validator(*m);
```
Comment (bb111189) [inline tests/Data/Test_Transaction.cpp]: Even though this is a test case, don't use new. Many times when  the team is trying to detect memory leak, we will run valgrind against unit test. By doing so, valgrind will flag it out as leak, even though in reality, the core code isn't

File: src/libValidator/Validator.cpp (hunk 7 above)
Comment (bb111189) [reply src/libValidator/Validator.cpp]: resolved

File: src/libValidator/Validator.cpp (hunk 6 above)
Comment (bb111189) [reply src/libValidator/Validator.cpp]: resolved

File: src/libValidator/Validator.cpp (hunk 5 above)
Comment (bb111189) [reply src/libValidator/Validator.cpp]: resolved

File: src/libValidator/Validator.cpp (hunk 4 above)
Comment (bb111189) [reply src/libValidator/Validator.cpp]: resolved. no changed required

File: src/libValidator/Validator.cpp (hunk 3 above)
Comment (bb111189) [reply src/libValidator/Validator.cpp]: resolved

File: src/libValidator/Validator.cpp (hunk 2 above)
Comment (bb111189) [reply src/libValidator/Validator.cpp]: resolved

**Comment by bb111189** [comment]
Other than the final comment above, I am ok with the pr

**Pull request merged**